Refactorization of Structure Generation Code, matching OO flow of FitSNAP

mitwood opened the pull request:
This will be a draft while large changes are collected. The goal will be to break out all key functions of the code to be callable classes and functions to avoid duplication and tangled information flow.

Commit:
Modifing the folder structure and putting placeholders for key
functions.

Commit:
Setting up input files in the form similar to fitsnap

Commit:
Renaming folders
	renamed:    lib/Cr-Fe_bcc_4.db -> GRSlib/Ver0_Files/Cr-Fe_bcc_4.db
	renamed:    lib/crystal_enum.py -> GRSlib/Ver0_Files/crystal_enum.py
	renamed:    lib/default_ACE_settings.py -> GRSlib/Ver0_Files/default_ACE_settings.py
	renamed:    lib/define_corr_basis.py -> GRSlib/Ver0_Files/define_corr_basis.py
	renamed:    lib/ga_tools.py -> GRSlib/Ver0_Files/ga_tools.py
	renamed:    lib/get_target_features.py -> GRSlib/Ver0_Files/get_target_features.py
	renamed:    lib/hnf.py -> GRSlib/Ver0_Files/hnf.py
	renamed:    lib/lammps_ACE_basis.py -> GRSlib/Ver0_Files/lammps_ACE_basis.py
	renamed:    lib/opt_tools.py -> GRSlib/Ver0_Files/opt_tools.py
	renamed:    lib/basis/construct.py -> GRSlib/basis/construct.py
	renamed:    lib/construct/ase_atoms.py -> GRSlib/construct/ase_atoms.py
	renamed:    lib/construct/base.py -> GRSlib/construct/base.py
	renamed:    lib/construct/get_target.py -> GRSlib/construct/get_target.py
	renamed:    lib/construct/lammps_data.py -> GRSlib/construct/lammps_data.py
	renamed:    lib/io/input.py -> GRSlib/io/input.py
	renamed:    lib/io/sections/basis.py -> GRSlib/io/sections/basis.py
	renamed:    lib/io/sections/constraint.py -> GRSlib/io/sections/constraint.py
	renamed:    lib/io/sections/error.py -> GRSlib/io/sections/error.py
	renamed:    lib/io/sections/genetic.py -> GRSlib/io/sections/genetic.py
	renamed:    lib/io/sections/motion.py -> GRSlib/io/sections/motion.py
	renamed:    lib/io/sections/output.py -> GRSlib/io/sections/output.py
	renamed:    lib/io/sections/scoring.py -> GRSlib/io/sections/scoring.py
	renamed:    lib/io/sections/sections.py -> GRSlib/io/sections/sections.py
	renamed:    lib/io/sections/target.py -> GRSlib/io/sections/target.py
	renamed:    lib/motion/genetic.py -> GRSlib/motion/genetic.py
	renamed:    lib/motion/gradient.py -> GRSlib/motion/gradient.py
	renamed:    lib/scoring/full_dist.py -> GRSlib/scoring/full_dist.py
	renamed:    lib/scoring/moments.py -> GRSlib/scoring/moments.py
	renamed:    lib/scoring/normalize.py -> GRSlib/scoring/normalize.py
	modified:   GenericInput.in
	deleted:    lib/__pycache__/crystal_enum.cpython-310.pyc
	deleted:    lib/__pycache__/default_ACE_settings.cpython-310.pyc
	deleted:    lib/__pycache__/ga_tools.cpython-310.pyc
	deleted:    lib/__pycache__/hnf.cpython-310.pyc
	deleted:    lib/__pycache__/lammps_ACE_basis.cpython-310.pyc
	deleted:    lib/__pycache__/opt_tools.cpython-310.pyc

Commit:
Moving over key functions from FitSNAP, adapting for GRS
	new file:   GRS.py
	new file:   __init__.py
	new file:   initialize.py
	new file:   parallel_output.py
	new file:   parallel_tools.py

Commit:
Adding basic code structures
	modified:   GRSlib/GRS.py
	modified:   GRSlib/io/sections/basis.py
	modified:   GRSlib/io/sections/scoring.py
	modified:   GenericInput.in

Commit:
Linking startup routines to input file needs
	modified:   ../GRSlib/GRS.py
	renamed:    ../GRSlib/construct/ase_atoms.py -> ../GRSlib/convert/ase_atoms.py
	renamed:    ../GRSlib/construct/base.py -> ../GRSlib/convert/base.py
	renamed:    ../GRSlib/construct/get_target.py -> ../GRSlib/convert/get_target.py
	renamed:    ../GRSlib/construct/lammps_data.py -> ../GRSlib/convert/lammps_data.py
	modified:   ../GRSlib/io/input.py
	modified:   ../GRSlib/io/sections/basis.py
	modified:   ../GRSlib/io/sections/constraint.py
	modified:   ../GRSlib/io/sections/genetic.py
	modified:   ../GRSlib/io/sections/motion.py
	modified:   ../GRSlib/io/sections/output.py
	modified:   ../GRSlib/io/sections/scoring.py
	new file:   ../GRSlib/io/sections/section_factory.py
	modified:   ../GRSlib/io/sections/sections.py
	modified:   ../GRSlib/io/sections/target.py

Commit:
IO class seems to be working, moving to core functions now. The template
for usage is the library mode of FitSNAP, see testing folder
	modified:   GRSlib/GRS.py
	modified:   GRSlib/io/input.py
	new file:   testing/FS_HONC-ACE.tmp
	new file:   testing/testing.py

Commit:
Adding more core functions to GRS.py
	modified:   GRSlib/GRS.py
	modified:   testing/testing.py

Commit:
Working on converter class (xyz)->(D)
	modified:   GRSlib/GRS.py
	deleted:    GRSlib/convert/lammps_data.py
	renamed:    GRSlib/convert/ase_atoms.py -> GRSlib/converters/ase_atoms.py
	renamed:    GRSlib/convert/base.py -> GRSlib/converters/base.py
	new file:   GRSlib/converters/convert.py
	new file:   GRSlib/converters/convert_factory.py
	renamed:    GRSlib/convert/get_target.py -> GRSlib/converters/get_target.py
	new file:   GRSlib/converters/lammps_base.py
	new file:   GRSlib/converters/lammps_pace.py
	modified:   GenericInput.in

Commit:
Forgot to modify variable names in a few io sections
	modified:   GRSlib/GRS.py
	modified:   GRSlib/io/sections/constraint.py
	modified:   GRSlib/io/sections/genetic.py
	modified:   GRSlib/io/sections/motion.py
	modified:   GRSlib/io/sections/scoring.py
	modified:   GRSlib/io/sections/target.py
	modified:   testing/testing.py

Commit:
Commited cache files, my bad
	deleted:    __pycache__/GRS.cpython-310.pyc
	deleted:    __pycache__/__init__.cpython-310.pyc
	deleted:    __pycache__/parallel_tools.cpython-310.pyc
	deleted:    io/__pycache__/input.cpython-310.pyc
	deleted:    io/sections/__pycache__/basis.cpython-310.pyc
	deleted:    io/sections/__pycache__/constraint.cpython-310.pyc
	deleted:    io/sections/__pycache__/error.cpython-310.pyc
	deleted:    io/sections/__pycache__/genetic.cpython-310.pyc
	deleted:    io/sections/__pycache__/motion.cpython-310.pyc
	deleted:    io/sections/__pycache__/output.cpython-310.pyc
	deleted:    io/sections/__pycache__/scoring.cpython-310.pyc
	deleted:    io/sections/__pycache__/section_factory.cpython-310.pyc
	deleted:    io/sections/__pycache__/sections.cpython-310.pyc
	deleted:    io/sections/__pycache__/target.cpython-310.pyc

jmgoff commented on GRSlib/basis/construct.py:
```diff
@@ -0,0 +1,71 @@
+        # radial character of fingerprints
+        #nmax = [3,2,1]
+    try:
+        from fitsnap3lib.lib.sym_ACE.yamlpace_tools.potential import AcePot
```
The more I think about it, the more I like importing the sym_ACE library from FitSNAP. That way, if we make any updates in the sym_ACE lib in FitSNAP, they are automatically used here. Otherwise, we will have to update both when we add/update anything for ACE.

jmgoff commented on GRSlib/converters/base.py:
```diff
@@ -0,0 +1,288 @@
+]
+
+
+elem_list = [
```
Looks like I left in some hard-coded lists of element types. We can remove this later.

jmgoff commented on GRSlib/io/sections/genetic.py:
It may be better to have a class above 'Genetic',  with something like "GlobalOptimization', that allows the user to do simulated annealing or a genetic algorithm. The simulated annealing may be better if people are looking at sampling composition space while freezing atomic positions. 

This can be changed later.

mitwood replied:
Are you thinking that the Optimizer class would then contain both genetic and gradient based moves?

jmgoff commented on GRSlib/io/sections/target.py:
```diff
@@ -0,0 +1,13 @@
+        self.allowedkeys = ['avg_weight', 'var_weight']
+        self._check_section()
+        self.target_fname = self.get_value("TARGET", "target_fname", None)
+        self.start_fname = self.get_value("TARGET", "start_fname", None)
```
`start_fname` may not be necessary in this class. In other scripts, `start_file` was a (probably poorly named) variable for the GRS or initial structure for which the descriptors needed to be calculated and compared to the target. However, they probably wont be needed to define the target itself.

mitwood replied:
I viewed the starting structure as the thing to be moved/mutated into the target descriptors. So this was one of my main tests, could you start from FCC/BCC and make changes such that you would get to the descriptors of BCC/FCC. Thinking about it more, the starting structure is more of a constraint than a target. Can move if you agree.

jmgoff reviewed (approved):
Thank you for the hard work on this! This is a good framework for the refactor. The legacy code is still there and can be used if needed. I didn't read through every change, but I think it would be better to merge this sooner rather than later. 

I only made a couple quick comments, but they don't need to be addressed before merging.

mitwood commented:
@jmgoff Mark anything you have looked at, Im still making a lot of changes since the only thing the code 'does' right now is make available the inputs as a dictionary. Converter is where I am currently, which is why there are some hard coded things.

Commit:
End of 3Mar25 updates
	new file:   ../.gitignore
	modified:   ../GRSlib/GRS.py
	renamed:    ../GRSlib/converters/base.py -> ../GRSlib/Ver0_Files/base.py
	new file:   ../GRSlib/converters/__init__.py
	modified:   ../GRSlib/converters/convert.py
	modified:   ../GRSlib/converters/convert_factory.py
	renamed:    ../GRSlib/converters/lammps_pace.py -> ../GRSlib/converters/lammps_ace.py
	modified:   ../GRSlib/converters/lammps_base.py
	modified:   ../GRSlib/io/input.py
	modified:   ../GRSlib/io/sections/basis.py
	modified:   ../GRSlib/io/sections/constraint.py
	modified:   ../GRSlib/io/sections/scoring.py
	modified:   ../GRSlib/io/sections/target.py
	modified:   ../GenericInput.in
	modified:   testing.py

Commit:
Working on scoring class to be consistent across all query types
	modified:   GRSlib/GRS.py
	deleted:    GRSlib/converters/ase_atoms.py
	modified:   GRSlib/converters/convert.py
	deleted:    GRSlib/converters/get_target.py
	modified:   GRSlib/converters/lammps_ace.py
	modified:   GRSlib/converters/lammps_base.py
	renamed:    GRSlib/motion/genetic.py -> GRSlib/motion/genetic/genetic.py
	renamed:    GRSlib/motion/gradient.py -> GRSlib/motion/gradient/gradient.py
	new file:   GRSlib/motion/moments.py
	new file:   GRSlib/motion/scoring.py
	deleted:    GRSlib/scoring/full_dist.py
	deleted:    GRSlib/scoring/moments.py
	deleted:    GRSlib/scoring/normalize.py
	modified:   GenericInput.in
	modified:   testing/testing.py

Commit:
Converters work, scoring class broken. Needs attn to JAX-LAMMPS coupling
	modified:   GRSlib/GRS.py
	modified:   GRSlib/converters/convert.py
	modified:   GRSlib/converters/convert_factory.py
	renamed:    GRSlib/converters/lammps_ace.py -> GRSlib/converters/sections/lammps_ace.py
	new file:   GRSlib/converters/sections/lammps_ase.py
	renamed:    GRSlib/converters/lammps_base.py -> GRSlib/converters/sections/lammps_base.py
	modified:   GRSlib/io/input.py
	modified:   GRSlib/io/sections/scoring.py
	new file:   GRSlib/motion/lossfunc/moments.py
	deleted:    GRSlib/motion/moments.py
	modified:   GRSlib/motion/scoring.py
	modified:   testing/testing.py

Commit:
End of 11Mar25 edits
	modified:   GRS.py
	modified:   io/sections/basis.py
	new file:   motion/genetic.py
	deleted:    motion/genetic/genetic.py
	new file:   motion/genetic_moves/alchemy.py
	new file:   motion/genetic_moves/density.py
	new file:   motion/genetic_moves/perturbation.py
	deleted:    motion/gradient/gradient.py
	modified:   motion/lossfunc/moments.py
	new file:   motion/motion.py
	modified:   motion/scoring.py

Commit:
Still working on genetic move, gradient looks ok for now. Scoring still
broken
	modified:   GRSlib/GRS.py
	modified:   GRSlib/motion/genetic.py
	modified:   GRSlib/motion/genetic_moves/alchemy.py
	modified:   GRSlib/motion/genetic_moves/density.py
	modified:   GRSlib/motion/genetic_moves/perturbation.py
	modified:   GRSlib/motion/motion.py
	modified:   GRSlib/motion/scoring.py
	modified:   GenericInput.in
	modified:   testing/testing.py

Commit:
Temporary add of testing folder
	new file:   testing/W54.lammps-data
	new file:   testing/bcc.data
	new file:   testing/coupling_coefficients.yace
	new file:   testing/log.lammps
	new file:   testing/notbcc.data
	new file:   testing/wig_LR_0_r123_lmax033.pickle

mitwood commented:
Closing this since it should be merged to develop before main.
